# Step 1: Deciphering the GPS coordinates

In [1]:
# secret_code = 9153787770964

secret_code = input("Enter encrypted message for the GPS Coordinates: ")

Enter encrypted message for the GPS Coordinates: 9153787770964


In [21]:
def decode_coordinates(secret_code):
  decoded_code = ""
  for digit in secret_code:
    shifted_digit = str((int(digit) - 4) % 10)  # Shift 4 places, wrap around at 10
    decoded_code += shifted_digit

  return decoded_code

In [22]:
decoded_coordinates = decode_coordinates(secret_code)
decoded_coordinates

'5719343336520'

In [23]:
len(decoded_coordinates)

13

In [24]:
# Latitude
lat_deg = decoded_coordinates[0]+decoded_coordinates[1]
lat_min = decoded_coordinates[2]+decoded_coordinates[3]
lat_sec = decoded_coordinates[4]+decoded_coordinates[5]
lat_dec = decoded_coordinates[6]

In [25]:
#Longitude
long_deg = decoded_coordinates[7]
long_min = decoded_coordinates[8]+decoded_coordinates[9]
long_sec = decoded_coordinates[10]+decoded_coordinates[11]
long_dec = decoded_coordinates[12]

In [26]:
print(f"The coordinates are:\n{lat_deg}°{lat_min}'{lat_sec}.{lat_dec}\"N {long_deg}°{long_min}'{long_sec}.{long_dec}\"W")

The coordinates are:
57°19'34.3"N 3°36'52.0"W


4 steps to left (-4):

Woodlands Terrace, Grantown-on-Spey, PH26 3JX, United Kingdom

The coordinates are:
57°19'34.3"N 3°36'52.0"W


------

4 steps to right (+4):

Mediterranean Sea

The coordinates are:
35°97'12.1"N 1°14'30.8"W

# Step 2: Deciphering the message on the scroll

N - modulus (3233)

E - exponent (17)

---

D = E^(-1)modφ(N)

φ(N) - totient function of N

---

N = 3233 = 61 x 53

p = 61

q = 53

---

φ(N) = (p-1)x(q-1) = 60 x 52 = 3120

In [10]:
E = 17
phi_N = 3120

In [11]:
from sympy import mod_inverse

In [12]:
D = mod_inverse(E, phi_N)

In [13]:
D

2753

## Cipher text decoding through RSA

Decryption - m(c) = c^d mod n

---

pow(x, y, z)

x	A number, the base

y	A number, the exponent

z	Optional. A number, the modulus

In [14]:
cipher_text = [2726, 1313, 1992, 884, 2412, 1453, 1230, 2185, 2412, 1992, 1313, 1230, 884, 1992, 281, 1632, 281, 2170, 1453, 1992, 1230, 2185, 2160, 1230, 1992, 745, 1632, 1992, 612, 745, 1632, 1627, 2160, 1313, 1992, 2412, 2185, 2160, 2923, 1313]

In [15]:
N = 3233

In [16]:
def decrypt_rsa(cipher_text, D, N):
  decrypted_text = [pow(c,D,N) for c in cipher_text]
  return decrypted_text

In [17]:
decrypted_numbers = decrypt_rsa(cipher_text, D, N)

In [18]:
decrypted_numbers

[76,
 101,
 32,
 116,
 114,
 233,
 115,
 111,
 114,
 32,
 101,
 115,
 116,
 32,
 99,
 97,
 99,
 104,
 233,
 32,
 115,
 111,
 117,
 115,
 32,
 108,
 97,
 32,
 112,
 108,
 97,
 113,
 117,
 101,
 32,
 114,
 111,
 117,
 103,
 101]

In [19]:
for alphabet in decrypted_numbers:
  print(chr(alphabet), end="")

Le trésor est caché sous la plaque rouge